modules/codegen/src/smithy4s/codegen/ToLine.scala

```diff
@@ -18,92 +18,160 @@ package smithy4s.codegen
-      case Primitive.BigInteger => line"BigInt"
-      case Primitive.Uuid       => Line(Set("java.util.UUID"), "UUID")
-      case Primitive.Document   => Line(Set("smithy4s.Document"), "Document")
-      case Primitive.Nothing    => line"Nothing"
+      case Primitive.Unit       => NameRef("Unit")
+      case Primitive.ByteArray  => NameRef("smithy4s", "ByteArray")
+      case Primitive.Bool       => NameRef("Boolean")
+      case Primitive.String     => NameRef("String")
+      case Primitive.Timestamp  => NameRef("smithy4s", "Timestamp")
+      case Primitive.Byte       => NameRef("Byte")
+      case Primitive.Int        => NameRef("Int")
+      case Primitive.Short      => NameRef("Short")
+      case Primitive.Long       => NameRef("Long")
+      case Primitive.Float      => NameRef("Float")
+      case Primitive.Double     => NameRef("Double")
+      case Primitive.BigDecimal => NameRef("BigDecimal")
+      case Primitive.BigInteger => NameRef("BigInt")
+      case Primitive.Uuid       => NameRef("java.util", "UUID")
+      case Primitive.Document   => NameRef("smithy4s", "Document")
+      case Primitive.Nothing    => NameRef("Nothing")
     }
 }
 
 // Models
-case class Line(imports: Set[String], line: String) {
-  def tupled = (imports, line)
-  def suffix(suffix: Line) = modify(s => s"$s $suffix")
-  def addImport(imp: String) = copy(imports = imports + imp)
-  def modify(f: String => String) = Line(imports, f(line))
-  def nonEmpty = line.nonEmpty
-  def toLines = Lines(imports, List(line))
-
-  def args(l: LinesWithValue*): Lines = if (l.exists(_.render.lines.nonEmpty)) {
-    val openBlock = if (line.nonEmpty) line + "(" else ""
-    Lines(imports, List(openBlock)) ++ indent(
-      l.toList.foldMap(_.render).mapLines(_ + ",")
-    ) ++ Lines(")")
-  } else Lines(line + "()").addImports(imports)
 
+sealed trait LineSegment { self =>
+  def toLine: Line = Line(Chain.one(self))
+}
+object LineSegment {
+  case class Import(value: String) extends LineSegment
+  object Import {
+    implicit val importShow = Show.show[Import](_.value)
+  }
+  case class Literal(value: String) extends LineSegment
+  object Literal {
+    implicit val literalShow = Show.show[Literal](_.value)
+  }
+  case class NameDef(name: String) extends LineSegment
+  object NameDef {
+    implicit val nameDefShow: Show[NameDef] = Show.show[NameDef](_.name)
+  }
+  case class NameRef(pkg: List[String], name: String) extends LineSegment {
+    self =>
+    def asValue: String = s"${(pkg :+ name).mkString(".")}"
+    def asImport: String = s"${(pkg :+ name.split("\\.")(0)).mkString(".")}"
+  }
+  object NameRef {
+    implicit val nameRefShow: Show[NameRef] = Show.show[NameRef](_.asImport)
+    def apply(pkg: String, name: String): NameRef =
+      NameRef(pkg.split("\\.").toList, name)
+    def apply(fqn: String): NameRef = {
+      val parts = fqn.split("\\.").toList
+      NameRef(parts.dropRight(1), parts.last)
+    }
+
+  }
+
```

Comment:
Might be nice to have some comments above the different line segment types to explain what they are and when they are used.